src/test/java/rx/internal/operators/OperatorValveTest.java

```diff
@@ -0,0 +1,163 @@
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+        // next batch
+        order.verify(inputProducer).request(2);// request limited called from onNext
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+        // next batch
+        order.verify(inputProducer).request(1);
+        control.onNext(false);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+
+        // no request while closed 
+        control.onNext(true);
+        control.onNext(true);
+
+        // small request while closed 
+        control.onNext(false);
+        outputProducer.request(1);
+        control.onNext(true);
+        order.verify(inputProducer).request(1);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+
+        // larger request while closed 
+        control.onNext(false);
+        outputProducer.request(2);
+        control.onNext(true);
+        order.verify(inputProducer).request(2);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+
+        // too large of a request while closed 
+        control.onNext(false);
+        outputProducer.request(3);
+        control.onNext(true);
+        order.verify(inputProducer).request(2);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+        order.verify(inputProducer).request(1);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+
+        input.onCompleted();
+        order.verify(output).onCompleted();
+
+        // all done
+        order.verifyNoMoreInteractions();
+    }
+
+    @Test
+    public void testDataError() {
+        TestSubscriber<Void> tSub = TestSubscriber.create();
+        Exception e = new Exception();
+        Observable.<Void> error(e).pressureValve(Observable.<Boolean> never(), 10).subscribe(tSub);
+        tSub.assertError(e);
+    }
+
+    @Test
+    public void testControlError() {
+        TestSubscriber<Void> tSub = TestSubscriber.create();
+        Exception e = new Exception();
+        Observable.<Void> never().pressureValve(Observable.<Boolean> error(e), 10).subscribe(tSub);
+        tSub.assertError(e);
+    }
+
+    @Test
+    public void testControlCompleteOpen() {
+        TestSubscriber<Integer> tSub = TestSubscriber.create();
+        Observable.just(1, 2, 3).pressureValve(Observable.<Boolean> empty(), 10).subscribe(tSub);
+        tSub.assertValues(1, 2, 3);
+    }
+
+    @Test
+    public void testControlCompleteClosed() {
+        TestSubscriber<Void> tSub = TestSubscriber.create();
+        Observable.<Void> never().pressureValve(Observable.just(false), 10).subscribe(tSub);
+        tSub.assertError(IllegalStateException.class);
+    }
+}
```

Comment:
Please write tests with the most common operators:

range.pressureValve.observeOn()
range.pressureValve.flatMap()
range.pressureValve.flatMap(maxConcurrent=1, 2, 4,...,128)
range.flatMap(v -> range.pressureValve)
range.pressureValve.concatMap()
range.concatMap(v -> range.pressureValve)